core/consensus/babe/impl/babe_impl.cpp

```diff
@@ -160,81 +162,95 @@ namespace kagome::consensus::babe {
       const primitives::BlockInfo announced_block(announce.header.number,
                                                   block_hash);
-
-      // Skip obsoleted announce
-      if (announced_block.number < current_best_block.number) {
-        return;
-      }
-
-      // Start catching up if gap recognized
-      if (announced_block.number > current_best_block.number + 1) {
-        startCatchUp(
-            peer_id, last_finalized_block, current_best_block, announced_block);
-      }
-
-      // Already has block with same number as ours best block or next of that
-      block_executor_->processNextBlock(
-          peer_id, announce.header, [this](const auto &header) {
-            onSynchronized();
-          });
+      startCatchUp(peer_id, announced_block);
+      return;
     }
+
+    // Already has block with same number as ours best block or next of that
+    babe_synchronizer_->enqueue(
+        announce.header,
+        peer_id,
+        [wp = weak_from_this()](
+            outcome::result<primitives::BlockInfo> block_res) {
+          if (auto self = wp.lock()) {
+            if (not block_res.has_value()) {
+              return;
+            }
+
+            if (self->current_state_ == Babe::State::CATCHING_UP) {
+              const auto &block = block_res.value();
+              SL_INFO(self->log_,
+                      "Catching up is finished on block #{} hash={}",
+                      block.number,
+                      block.hash.toHex());
+              self->current_state_ = Babe::State::SYNCHRONIZED;
+            }
+            self->onSynchronized();
+          }
+        });
   }
 
   void BabeImpl::startCatchUp(const libp2p::peer::PeerId &peer_id,
-                              const primitives::BlockInfo &finalized_block,
-                              const primitives::BlockInfo &best_block,
                               const primitives::BlockInfo &target_block) {
     // synchronize missing blocks with their bodies
     SL_INFO(log_, "Catching up to block #{} is ran", target_block.number);
     current_state_ = State::CATCHING_UP;
 
-    block_executor_->requestBlocks(
-        finalized_block.hash,
-        target_block.hash,
+    babe_synchronizer_->enqueue(
+        target_block,
         peer_id,
-        [wp = weak_from_this(), bn = target_block.number] {
+        [wp = weak_from_this(),
+         bn = target_block.number](outcome::result<primitives::BlockInfo> res) {
           if (auto self = wp.lock()) {
-            SL_INFO(self->log_, "Catching up to block #{} is done", bn);
-            self->current_state_ = State::WAIT_BLOCK_ANNOUNCE;
+            if (not res.has_value()) {
+              SL_INFO(self->log_,
+                      "Catching up to block #{} is failed: ",
+                      bn,
+                      res.error().message());
+              return;
+            }
+
+            SL_INFO(self->log_, "Catching up to block #{} in progress", bn);
```

Comment:
Hm, I'm a bit confused why when this enqueue invokes the callback catching up is in progress, but when enqeue from onBlockAnnounce does, it means catching up is done. My confusion must be caused by lack of knowledge in Babe, but still, maybe a comment here on this would make sense.

Reply:
Catching up is initiates here, but happens asynchronously